Adding a CSP Middleware

burzum opened the pull request:
This adds a first iteration of a CSP middleware based on https://github.com/paragonie/csp-builder

Be aware that this feature *requires* php 7. This has been discussed with @markstory before:

> [21:03] 
> Hmm the main problem I have is anyway the php7 requirement.
> [21:03] 
> Can we use it if we just suggest it?
> [21:06] 
> markstory I guess we could

Appveyor will fail on setup because it only features php 5.6. So ignore that CI service.

burzum committed:
Adding methods to the entity to check for values

burzum committed:
CS fix

ADmad committed:
Bump up phpstan to level 2.

motooka committed:
Add warning log when a connection with active transaction is going to be closed.

motooka committed:
Removed trailing whitespace

motooka committed:
Added test code for the warning log message content.

motooka committed:
Added condition to check Log class existence

motooka committed:
Merge branch 'master' into warning-log-on-active-transaction

# Conflicts:
#	src/Database/Connection.php

markstory committed:
Update version number to 3.5.0

markstory committed:
Update version number.

chinpei215 committed:
Remove meaningless code

chinpei215 committed:
Fix string handling of Hash::remove() and Hash::_simpleOp('remove')

Hash::remove() mishandles '0' in path strings.
Hash::_simpleOp('remove') causes a fatal error when data is a string.

Commit:
Fix up behavior callbacks for clean inheritance and consistency.

ADmad committed:
Fix errors reported by phpstan in Error package

ADmad committed:
Add annotations for missing methods in ConnectionInterface.

ADmad committed:
Add annotations for missing methods to interfaces.

This prevents static analyzers like phpstan complaining when an argument is
typehinted as interface and a method used on object is not present in the interface.

Closes #11051

ADmad committed:
Fix issues reported by phpstan in Database package.

ADmad committed:
Fix error reported by phpstan in I18n/RelativeTimeFormatter.

ADmad committed:
Avoid code duplication.

markstory committed:
Merge pull request #11066 from cakephp/i18n-translator

Avoid code duplication.

markstory committed:
Merge branch '3.next' into master-callbacks

chinpei215 committed:
Use is_array() isntead of !is_string()

ADmad committed:
Use where() instead of andWhere().

This avoid having to add annotation for andWhere() to QueryInterface.

markstory committed:
Merge pull request #11063 from cakephp/issue-11051

Add annotations for missing methods to interfaces.

ADmad committed:
Merge branch 'master' into phpstan-l-2

inoas committed:
Session $id is string|int

inoas committed:
More docblocks

inoas committed:
Session $id is string|int

ADmad committed:
Fix errors reported by phpstan in Database package.

inoas committed:
Update CacheSession.php

Add dot

ADmad committed:
Fix errors reported by phpstan in ORM package

ADmad committed:
Fix errors reported by phpstan in TimeHelper.

ADmad committed:
Ignore error reported by phpstan in FixtureInjector.

ADmad committed:
Add annoations to avoid ignoring phpstan reported errors

markstory committed:
Merge pull request #11068 from inoas/patch-7

3.x Docblock CacheSession $id is string|int

markstory committed:
Merge pull request #11069 from inoas/patch-9

3.x Docblock DatabaseSession $id is string|int

markstory committed:
Merge pull request #11060 from chinpei215/fix-hash-remove

Fix string handling of Hash::remove() and Hash::_simpleOp('remove')

chinpei215 committed:
Fix notMatching for BelongsToMany with composite keys

ADmad committed:
Fix phpstan reported error in Xml class.

ADmad committed:
Ignore incorrect errors reported for Redis class methods.

ADmad committed:
Add method annotations to avoid ignoring errors.

chinpei215 committed:
Fix parseDSN not working for SqlLocalDB

Also fix the regex for url with fragment.

ADmad committed:
Fix property type.

lorenzo committed:
Merge pull request #11073 from chinpei215/fix-not-matching

Fix notMatching for BelongsToMany with composite keys

markstory committed:
Merge pull request #11071 from chinpei215/fix-parse-dsn

Fix parseDSN not working for SqlLocalDB

markstory committed:
Add a deprecation warning helper.

This will make adding framework deprecation warnings we plan to add in
3.6.x simpler as they will be easier to write.

ADmad committed:
Revert back "andWhere()" calls.

chinpei215 committed:
Fix Collection throwing an exception when Traversable given.

lorenzo committed:
Merge pull request #11076 from chinpei215/fix-collection-travarsable

Fix Collection throwing an exception when Traversable given.

ADmad committed:
Update docblocks in StatementDecorator class.

markstory committed:
Fix failing test.

Test was failing because error_reporting level was modified by another
test.

ADmad committed:
Merge pull request #11031 from cakephp/phpstan-l-2

Bump up phpstan to level 2.

markstory committed:
Fix response headers set in templates not be sent.

When response properties are set in the view templates, they should show
up in the controller's response as well.

Fixes #11084

markstory committed:
Handle immutable responses better.

Wait until the afterDispatch event to grab the controller response
property. Event listeners in the shutdown or afterDispatch events may
have replaced the controller's response.

chinpei215 committed:
Fix parseDsn() ignoring empty strings

ADmad committed:
Fix CS error.

ceeram committed:
Escape urls to avoid xss

ceeram committed:
HTTP_X_FORWARDED_FOR can be spoofed, proxies append to the list, so use last ip

ceeram committed:
Change default for parseHuge to false to avoid OOM on xml expansion

markstory committed:
Merge pull request #11087 from cakephp/issue-11084

Fix response headers set in templates not be sent.

ceeram committed:
Ensure input passwords will be hashed even when a user doesn't exist

ceeram committed:
Use timing attack safe string comparison

markstory committed:
Merge pull request #11093 from cakephp/client-ip-spoofing

HTTP_X_FORWARDED_FOR can be spoofed, proxies append to the list, so u…

markstory committed:
Fix cookies not being sent by classic Dispatcher.

Cookies are not arrays anymore. We lacked test coverage for this in the
past.

Refs #11070

lorenzo committed:
Merge pull request #11097 from cakephp/issue-11070

Fix cookies not being sent by classic Dispatcher.

lorenzo committed:
Merge pull request #11096 from cakephp/digest-constant-time

Use timing attack safe string comparison

lorenzo committed:
Merge pull request #11095 from cakephp/time-based-attack

Ensure input passwords will be hashed even when a user doesn't exist

dereuromark committed:
Add iterator annotation

ceeram committed:
Do not fallback to insecure randomBytes, should be called explicitly

ceeram committed:
Do not serve hidden files or files in hidden folders

markstory committed:
Merge pull request #11098 from cakephp/master-annotation

Add iterator annotation

markstory committed:
Merge pull request #11094 from cakephp/xml-expansion

Change default for parseHuge to false to avoid OOM on xml expansion

markstory committed:
Merge pull request #11090 from chinpei215/fix-parse-dsn-regexp

Fix parseDsn() ignoring empty strings

markstory committed:
Merge pull request #11100 from cakephp/hidden-files-folders-middleware

Do not serve hidden files or files in hidden folders

ceeram committed:
Use hmac for token hashes to avoid collisions

ceeram committed:
Replace md5 with hmac hashes to provide unpredictability and resistance to collisions

ceeram committed:
Use hmac for digest nonces

ceeram committed:
update cacert.pem to latest version

markstory committed:
Add assertErrorEmpty()

This is handy for making sure a shell doesn't output any error messages.

markstory committed:
Move RoutesShell test to use the integration test case.

This new test case class lets us use less code to test shell commands.

markstory committed:
Merge pull request #11104 from cakephp/cacert

update cacert.pem to latest version

markstory committed:
Merge pull request #11102 from cakephp/autolink-hmac

Replace md5 with hmac hashes

markstory committed:
Add more assertion methods to ConsoleIntegrationTestCase.

Add assertions for checking table rows, and empty output.

markstory committed:
Add assertions for error codes to routes shell.

markstory committed:
Add stackframe information to deprecation message.

Having the userland method line/file should help in fixing deprecation
warnings. I've not used Debugger here as it would add a dependency to
cakephp/core that will not be easy to satisfy for standalone packages.

markstory committed:
Merge pull request #11105 from cakephp/routes-shell-test-updates

Routes shell test updates

markstory committed:
Add an alias for Deprecated to be compatible with PHP5.6

markstory committed:
Remove unset() and clean up tests.

Use mocks instead of writing to log files. Don't mock Connection, as it
is the class under test, and the mocked driver obviates the need to mock
connect().

markstory committed:
Emit log messages when connections are destroyed mid transaction

Merge branch 'issue-11032' into master.

Refs #11032

markstory committed:
Convert CacheShell to use ConsoleIntegrationTestCase

ADmad committed:
Add Form::setErrors().

Closes #10971

markstory committed:
Merge pull request #11103 from cakephp/digest-nonce-hmac

Use hmac for digest nonces

burzum committed:
Adding a test to show the translation issue of plurals

markstory committed:
Toggle deprecation warnings on error level.

By using the error level to control deprecation warnings, we allow users
to have them enabled in production but go into log files. This is useful
for ensuring all the deprecation warnings have been caught/fixed.

markstory committed:
Merge pull request #11108 from cakephp/cache-shell-test

Convert CacheShell to use ConsoleIntegrationTestCase

burzum committed:
Adding a CSP Middleware

burzum committed:
Adding a CSP Middleware

ADmad commented on src/Http/Middleware/CspMiddleWare.php:
```diff
@@ -0,0 +1,80 @@
+    /**
+     * Constructor
+     *
+     * @param null|\ParagonIE\CSPBuilder\
```
Shouldn't this include `array` too?

burzum replied:
Yes, I've had the idea of supporting an array after I wrote that. ;)

burzum committed:
Skip the CSP tests if php lower than 7.0

burzum committed:
Skip the CSP tests if php lower than 7.0

ADmad commented on tests/TestCase/Http/Middleware/CspMiddlewareTest.php:
```diff
@@ -0,0 +1,68 @@
+            return $response;
+        };
+
+        $middleware = new CspMiddleware([
```
Need a test passing `\ParagonIE\CSPBuilder\` instance too to the constructor.

burzum replied:
Yes, I'll wait with further changes until I've got more feedback if the approach of the implementation will be accepted.

markstory committed:
Merge pull request #11092 from cakephp/html-helper-escape-urls

Escape urls to avoid xss

markstory committed:
Merge pull request #11109 from cakephp/form-set-errors

Add Form::setErrors().

markstory commented on composer.json:
```diff
@@ -38,11 +38,13 @@
         "phpunit/phpunit": "^5.7|^6.0",
-        "cakephp/cakephp-codesniffer": "^3.0"
+        "cakephp/cakephp-codesniffer": "^3.0",
+        "paragonie/csp-builder": "^2.1"
```
Won't this break builds on php5 going forwards? Could we conditionally include this dependency in travis instead of breaking the php5 builds?

burzum replied:
We can do that. Any idea how? I'm not really good with the scripts for Travis and don't want to produce 10 or 20 trial and error commits just to get it working.

markstory commented on src/Http/Middleware/CspMiddleWare.php:
```diff
@@ -0,0 +1,80 @@
+    public function __construct($csp = null)
+    {
+        if (!is_array($csp)) {
+            $cspConfig = (array)Configure::read('App.CSP');
```
Why would this be an application config value? Do you see the CSP header varying by environment?

ADmad replied:
In general too let's not make new classes directly read from `Configure::read()` if possible. Even if someone is saving the setting in `Configure` they can easily do `new CspMiddleware(Configure::read('App.CSP'))`.

burzum replied:
It's not creating a new class based on it, it's just passing the configuration to the new instance.

ADmad replied:
There's no problem using `Configure` to store configure. But there's no need to use `Configure::read()` inside the class constructor where you can easily read and pass the config array as argument.

For datasouces also we do the [same](https://github.com/cakephp/app/blob/master/config/bootstrap.php#L146). Read the config array and pass it to `ConnectionManager::setConfig()`.

burzum replied:
Ah OK, I get your point now. But the problem is then, even if we provide defaults, the user would still have to be aware of them and manually pass them. I was looking for a convenient way to make CSP work out of the box with some good default settings.

markstory replied:
We can make it work out of the box by updating the application skeleton as needed.

markstory committed:
Update version number to 3.5.1

markstory committed:
Update documentation to be clearer on how marshalling works

Document that setters will not be triggered, and fields will not be
marked dirty if the value is an identical scalar and an entity is
updated.

ADmad committed:
Fix filename

ADmad committed:
Fix filename

ADmad committed:
Add missing "use" statement.

ADmad committed:
Add missing "use" statement.

ADmad committed:
Use paragonie/csp-builder only for PHP 7.x.

ADmad committed:
Use paragonie/csp-builder only for PHP 7.x.

ADmad commented:
I have made a couple of fixes and update travis config to load `paragonie/csp-builder` only for PHP 7.x.

burzum committed:
Adding another test case for the CspMiddleware

burzum committed:
Adding another test case for the CspMiddleware

markstory committed:
Address feedback from review.

burzum committed:
phpcs fixes

burzum committed:
phpcs fixes

dereuromark commented on src/Http/Middleware/CspMiddleware.php:
```diff
@@ -36,7 +36,7 @@ class CspMiddleware
      * Constructor
      *
-     * @param null|\ParagonIE\CSPBuilder\
+     * @param null|\ParagonIE\CSPBuilder\ $csp CSP object or config array
```
The trailing \ is too much

markstory committed:
Merge pull request #11117 from cakephp/document-setter-behavior

Update documentation to be clearer on how marshalling works

markstory committed:
Windows path fix.

supermanner committed:
Fix CookieCollection::parseSetCookieHeader when the cookie has properties that are not expected in Cookie Object

saeideng committed:
fix version

markstory committed:
Merge pull request #11123 from saeideng/fix-version

fix version

supermanner committed:
Add test to check whether unexpected cookie propery isn't included

markstory committed:
Merge pull request #11062 from cakephp/master-callbacks

Fix up behavior callbacks for clean inheritance and consistency.

markstory committed:
Merge pull request #11121 from supermanner/fix-cookie-collection

Fix CookieCollection::parseSetCookieHeader when the cookie has proper…

markstory committed:
Add casting to make CookieComponent more backwards compatible.

Add casting that allows CookieComponent to continue working with loose
typing that was supported in earlier versions.

Refs #11120

markstory committed:
Don't replace plural message ids with singular ones.

When a messageid is defined as a plural key, and *then* a singular one
in the same context, the singular only message id should be ignored as
message selection methods will choose the first value from the plural
values. The plural translations obviously cannot read multiple values
from the singular data, causing missed messages.

Refs #11089

ADmad committed:
Update docblocks.

huwcbjones committed:
Fix #11132 - trim warning in Cake\View\Helper::addClass

markstory committed:
Merge pull request #11130 from cakephp/docblock

Update docblocks.

huwcbjones committed:
Added isset.
Fix coding style

ADmad committed:
Make View's setter methods return $this instead of void.

markstory committed:
Merge pull request #11127 from cakephp/3.next-i18n-plural-issue

Fix plural translations falling back when they shouldn't

ADmad committed:
Add tests.

huwcbjones committed:
Added test case for addClass()

huwcbjones committed:
Short array syntax

huwcbjones committed:
Removed blank line

markstory committed:
Fix redis increment/decrement with infinite timeout.

When timeout is 0 (live forever) no TTL should be set on
incremented/decremented keys.

Refs #11135

markstory committed:
Merge pull request #11134 from cakephp/view-retun-this

Make View's setter methods return $this instead of void.

markstory committed:
Merge pull request #11075 from cakephp/36-deprecate-function

3.6 - Add a deprecation warning helper.

markstory committed:
Merge branch 'master' into 3.next

markstory committed:
Merge branch '3.next' of github.com:cakephp/cakephp into 3.next

markstory committed:
Merge pull request #11133 from huwcbjones/view_helper_trim

Fix #11132 - trim warning in Cake\View\Helper::addClass

markstory committed:
Tidy up.

Refs #11133

markstory committed:
Merge pull request #11099 from cakephp/no-insecure-fallback

Do not fallback to insecure randomBytes, should be called explicitly

markstory committed:
Add more context on formhelper templates.

Add more context/usage for each template string that FormHelper uses.

Refs cakephp/docs#5232

lorenzo committed:
Merge pull request #11139 from cakephp/issue-5232

Add more context on formhelper templates.

huwcbjones committed:
Added support for `fullBase` option in `::link()`.
Added test case for `fullBase`.

markstory committed:
Merge pull request #11126 from cakephp/issue-11120

Add casting to make CookieComponent more backwards compatible.

markstory committed:
Merge pull request #11136 from cakephp/redis-increment-3x

Fix redis increment/decrement with infinite timeout.

markstory committed:
Merge branch 'master' into 3.next

markstory committed:
Improve warnings around Text::uuid()

This function should not be used to generate 'secure' identifiers. While
UUID v4 has a low chance of collisions it is not as foolproof as 64
fully random bytes.

I've also opted to use a more secure source of random integers in PHP7
as mersenne twister to reduce the risk of if these uuids are used in
a security related context.

huwcbjones committed:
Changed array_key_diff to unset

inoas committed:
Bring gitattributes in line with cakephp/app

...regarding binary vs text treatment, eol/crlf/lf

markstory committed:
Merge pull request #11146 from inoas/patch-10

Bring gitattributes in line with cakephp/app

markstory committed:
Update CommandListShell to use integration test case.

markstory committed:
Update HelpShell tests to use ConsoleIntegrationTestCase

ceeram committed:
Add user specifc data to token hash to avoid privilige escalation

markstory committed:
Merge pull request #11151 from cakephp/users-token

Add user specifc data to token hash to avoid privilige escalation

Commit:
Fix error output line highlighting off by one.

The debugger highlights the line before the actual line of error.

markstory committed:
Merge pull request #11153 from ndm2/3.x-fix-debugger-line-highlighting

3.x - Fix error output line highlighting off by one.

burzum committed:
Fixing doc block param declarations

burzum committed:
Fixing doc block param declarations

burzum committed:
Fixing php doc blocks

burzum committed:
Fixing php doc blocks

markstory committed:
Merge pull request #11150 from cakephp/console-test-update

Update 2 more shell tests to use ConsoleIntegrationTestCase

dereuromark committed:
Fix CLI session ID to be set, resolves regression.

batopa committed:
load routes in RoutingMiddleware just if not initialized

markstory committed:
Merge pull request #11161 from batopa/issue/11160-fix-integration-tests-using-named-routes

Fix integration tests when named routes are used

markstory committed:
Merge pull request #11142 from cakephp/uuid-warning

Improve warnings around Text::uuid()

dereuromark committed:
Alternative approach to session id in CLI

markstory committed:
Merge pull request #11158 from cakephp/dereuromark-patch-1

Fix CLI session ID to be set, resolves regression.

robertpustulka committed:
Support protocol relative URL in HTTP Client.

Some servers set //example.com as a Location header. `buildUrl()` needs
to support that in order to build absolute urls.

robertpustulka committed:
Require that cookies added to the Client have a path and a domain.

robertpustulka committed:
Fix adding request cookies for urls without a path.

markstory committed:
Merge pull request #11164 from robertpustulka/restrict-client-cookies

Require that cookies added to the Client have a path and a domain.

markstory committed:
Merge pull request #11167 from robertpustulka/fix-add-to-request

Fix adding request cookies for urls without a path.

markstory committed:
Merge pull request #11141 from huwcbjones/ticket-11140

Fix #11140 - Added support for `fullBase` option in `::link()`.

Commit:
Mute E_DEPRECATED warnings during tests and add depreciation warnings to TableSchema

markstory committed:
Merge pull request #11101 from cakephp/token-hmac

Use hmac for token hashes to avoid collisions

markstory committed:
Update version number to 3.5.2

markstory committed:
Use numeric constant value.

We can't use the constant expression in the XML file.

markstory committed:
Fix deprecated API usage for TableSchema.

antograssiot committed:
Remove more deprecated method from TableSchema

markstory committed:
Fix backwards compatibility with CommandRunner.

In making the new runner stricter, I dropped the inflection on command
names.

Refs #11174

Commit:
Update the typehint to Connection instead of ConnectionInterface

robertpustulka committed:
Added `protocolRelative` option to `buildUrl()` method.

littleylv committed:
Display help message with `out` instead of `err` when missing method

markstory committed:
Merge pull request #11175 from cakephp/issue-11174

Fix backwards compatibility with CommandRunner.

markstory committed:
Merge pull request #11171 from cakephp/add-deprecation-warnings

Mute E_DEPRECATED warnings during tests and add depreciation warnings

dereuromark committed:
Fix annotation for main()

markstory committed:
Merge pull request #11180 from cakephp/dereuromark-patch-1

Fix annotation for main()

markstory committed:
Merge pull request #11177 from cakephp/update-schema-cache-typehint

Update the typehint to Connection instead of ConnectionInterface

littleylv committed:
Show error message before output help when missing method

markstory committed:
Merge pull request #11179 from littleylv/missing-method

Display shell help message with `out` instead of `err` when missing method

markstory committed:
Merge pull request #11163 from robertpustulka/schemeless-url-fix

Support protocol relative URL in HTTP Client.

markstory committed:
Merge branch 'master' into 3.next

markstory committed:
Fix SecurityComponent tests.

They were not updated to account for the session id being added to the
checksum.

markstory committed:
Fixate the session_id to avoid having to update all the tokens.

markstory committed:
Remove hasProperty() & update doc blocks.

This method was an alias of has(). I'm opting to keep `has` for a few
reasons:

* It is shorter
* It is in the existing interface. Backwards compat win.
* It does what it says it does.

markstory committed:
Merge branch '3.next-entity-methods' into 3.next

burzum commented:
@markstory @ADmad what is left here? The two tests that fail are failing because of something related to the connection interface which I haven't even touched at all in this PR.

cleptric commented:
@burzum I fixed this in another PR. So just rebase 3.next into your branch or someone else could just merge it 🙂

burzum committed:
Merge branch '3.next-csp' of github.com:cakephp/cakephp into 3.next-csp
